### PR TITLE
D graph

### PR DESCRIPTION
**Pull request by ksgifford**

### CODE REVIEW

File: src/d_graph.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        """Return list of keys in the dict, which represent node names."""
+        return list(self._d_graph.keys())
+
+    def edges(self):
+        """Return a list of edges that originate from each node."""
+        edge_list = []
+        for key, val in self._d_graph.items():
+            for node in val:
+                edge_list.append('>'.join((key, node)))
+        return edge_list
```
Comment (nhuntwalker) [inline src/d_graph.py:22]: You could condense this to one line with list comprehension

File: src/d_graph.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    def del_node(self, node):
+        """Remove specified node from graph."""
+        try:
+            del self._d_graph[node]
+            for k, v in self._d_graph.items():
+                try:
+                    v.remove(node)
+                except KeyError:
+                    pass
```
Comment (nhuntwalker) [inline src/d_graph.py:41]: Instead of handling an exception, you can just remove the node if the node is in the node list. A conditional bypasses the need to deal with an error.

File: src/test_dgraph.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    test_graph.add_edge('A', 'B')
+    test_graph.add_edge('A', 'D')
+    test_graph.add_edge('A', 'C')
+    test_graph.add_edge('B', 'C')
+    return test_graph
+
+
+def test_nodes(instance):
+    assert sorted(instance.nodes()) == ['A', 'B', 'C', 'D']
+
```
Comment (nhuntwalker) [inline src/test_dgraph.py:23]: Should also test that if there are no nodes, an empty list is returned

File: src/test_dgraph.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+def test_edges(instance):
+    expected = ['A>B', 'A>C', 'A>D', 'B>C']
+    assert sorted(instance.edges()) == sorted(expected)
+
```
Comment (nhuntwalker) [inline src/test_dgraph.py:28]: Should also test that a graph with no edges returns an empty list.

File: src/test_dgraph.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+def test_add_nodes(instance):
+    instance.add_node('E')
+    assert sorted(instance.nodes()) == ['A', 'B', 'C', 'D', 'E']
+
+
+def test_add_edge(instance):
+    instance.add_edge('C', 'D')
+    expected = ['A>B', 'A>C', 'A>D', 'B>C', 'C>D']
+    assert sorted(instance.edges()) == sorted(expected)
+
```
Comment (nhuntwalker) [inline src/test_dgraph.py:39]: this also needs to test that nodes "C" and "D" were added if they didn't already exist in the graph
